### PR TITLE
New customPath parameter with test

### PR DESCRIPTION
**Pull request by laurentschall**
Hi Invertase team,

This is the same as [https://github.com/invertase/dart_custom_lint/pull/187](url), but including a test on the new **customPath** parameters.

Please let me know if this test is sufficient. I added no test for the other methods where the **customPath** parameter has been integrated as well in the pull request of Matthias, because I did not found any existing test to extend for those other methods. I even found no call at all to the addYamlFileEdit() method.

Thanks in advance for the review.

Best Regards,
Laurent

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -9,9 +9,6 @@ on:
       - main
     paths-ignore:
       - "**.md"
-  schedule:
```
Comment (rrousselGit) [inline .github/workflows/build.yml:12]: This file shouldn't be edited

File: packages/custom_lint_core/test/assist_test.dart [hunk 2]
```diff
@@ -58,6 +64,36 @@ void main() {
...
+''');
+    final result = await resolveFile2(path: file.path);
+    result as ResolvedUnitResult;
+
+    final changeList1 = await assist1.testRun(result, SourceRange.EMPTY);
+    final changeList2 = await assist2.testRun(result, SourceRange.EMPTY);
+
+    expect(
+      changeList1,
+      matcherNormalizedPrioritizedSourceChangeSnapshot('custom.json'),
```
Comment (rrousselGit) [inline packages/custom_lint_core/test/assist_test.dart:84]: We don't care about normalized snapshots, since they don't include file paths – which is exactly what we care about here.

File: packages/custom_lint_core/test/assist_test.dart [hunk 3]
```diff
@@ -58,6 +64,36 @@ void main() {
...
+    );
+    expect(jsonEncode(changeList1[0]).contains('custom_1.txt'), true);
```
Comment (rrousselGit) [inline packages/custom_lint_core/test/assist_test.dart:86]: Don't jsonEncode the object and instead read the map property corresponding to the file path

Comment (laurentschall) [reply packages/custom_lint_core/test/assist_test.dart]: I added **_extractFileName()** method that reads the nested maps to extract the file path value.

File: packages/custom_lint_core/test/assist_test.dart (hunk 2 above)
Comment (laurentschall) [reply packages/custom_lint_core/test/assist_test.dart]: Then I removed this expect() statement and the subsequent code and test files.

File: .github/workflows/build.yml (hunk 1 above)
Comment (laurentschall) [reply .github/workflows/build.yml]: Now the file is not included in the pull request anymore.

Comment (laurentschall) [reply .github/workflows/build.yml]: I randomly have the following error in (master,get). One day it occurs and the other day not. But I see the same kind of error in the build of the original branch so I guess this is not related to this pull request content.

 _[custom_lint]: ::group::❌ test/server_test.dart: hot-reload handles the source change of one plugin and restart it (failed)
[custom_lint]: **PathNotFoundException**: Failed to watch path, path = '/home/runner/work/dart_custom_lint/dart_custom_lint/packages/custom_lint/.dart_tool/**custom_lint_testBZZXHU' (OS Error: No such file or directory**, errno = 2)_

**Comment by rrousselGit** [comment]
Sorry this flew under the radar.
LGTM, thanks!

**Comment by laurentschall** [comment]
You're welcome, thank you for the merge !

**Comment by ldsunit** [comment]
tldr; `File(customPath).readAsStringSync()` returns the original file contents even after a fix is applied.

I think this is problematic because the file doesn't get saved after making modifications. I'm trying to add entries to a json file. The first fix works fine, but subsequent fixes are still referencing an old version of the file. The changeBuilder seems to know about the latest file version, but I don't see a way to access the `workspace` to read the file. 

A simplified version of this is:

```dart
final path = 'some/file/path.json';

// read file contents (on subsequent fixes, this returns the version before the first fix)
final fileContents = File(path).readAsStringSync();
final asJson = jsonDecode(fileContents) as Map<String, dynamic>;
final encoder = JsonEncoder.withIndent('  ');

if (!asJson.containsKey('foo')) {
  changeBuilder.addGenericFileEdit(
    (builder) {
      final lastBracketPosition = fileContents.lastIndexOf('}');

      builder.addSimpleReplacement(
        SourceRange(lastBracketPosition - 1, 2),
        ',\n  "foo": "bar"',
      );
    },
    customPath: path,
  );
}
```

if `some/file.path.json` starts as 

```json
{
  "one": "two"
}
```

applying a fix the first time results in:

```json
{
  "one": "two",
  "foo": "bar",
}
```

applying the fix again results in:

```json
{
  "one", "two",
  "foo": "bar"
}  "foo": "bar"
}
```

I noticed that the analyzer plugin [gets file contents](https://github.com/dart-lang/sdk/blob/main/pkg/analyzer_plugin/lib/src/utilities/change_builder/change_builder_core.dart#L222-L224) from `ChangeWorkspace`, but I don't see how to access that.

**Comment by unitmdev** [comment]
opened issue #296
